File: Python3/201 - 400/387 First Unique Character in a String.py
# 80ms, 90.29%
class Solution:
    def firstUniqChar(self, s):
        alphabet = 'abcdefghijklmnopqrstuvwxyz'
        res = [s.index(c) for c in alphabet if s.count(c) == 1]
        return min(res) if res else -1
    '''Time complexity
    above: T(n) ~ O(26n), costs 80ms
    below: T(n) = O(3n), costs 160ms ... WHY?
    '''
        x = [0] * 26
        s_int = [ord(c) - 97 for c in s]
        s_end = len(s_int)
        for c in s_int:
            x[c] += 1
        for i in range(s_end):
            if x[s_int[i]] == 1: return i
        return -1
    

File: 201 - 400/387 First Unique Character in a String.cc
// 56ms, 59.12%
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int firstUniqChar(string &s) const {
        int len = s.size();
        if (len) {
            if (len == 1) return 0;
            int count[26] {};
            char *s_ptr = &s[0], *s_end = s_ptr + len, *p = s_ptr;
            while (p < s_end) ++count[*p++ - 'a'];
            p = s_ptr;
            while (p < s_end) {
                if (count[*p - 'a'] == 1)
                    return int(p - s_ptr);
                ++p;
            }
        }
        return -1;
    }
};
